packages/flutter/lib/src/widgets/focus_scope.dart

```diff
@@ -1,145 +1,343 @@
+import 'package:flutter/foundation.dart';
+import 'package:flutter/services.dart';
 
 import 'basic.dart';
 import 'focus_manager.dart';
 import 'framework.dart';
+import 'raw_keyboard_listener.dart';
 
-class _FocusScopeMarker extends InheritedWidget {
-  const _FocusScopeMarker({
-    Key key,
-    @required this.node,
-    Widget child,
-  }) : assert(node != null),
-       super(key: key, child: child);
-
-  final FocusScopeNode node;
+typedef FocusOnKeyCallback = bool Function(FocusNode node, RawKeyEvent event);
 
-  @override
-  bool updateShouldNotify(_FocusScopeMarker oldWidget) {
-    return node != oldWidget.node;
-  }
-}
-
-/// Establishes a scope in which widgets can receive focus.
+/// A widget that manages a [FocusNode] to allow keyboard focus to be given
+/// to this widget and its descendants.
 ///
-/// The focus tree keeps track of which widget is the user's current focus. The
-/// focused widget often listens for keyboard events.
+/// It manages a [FocusNode], managing its lifecycle, and listening for
+/// changes in focus.
 ///
-/// A focus scope does not itself receive focus but instead helps remember
-/// previous focus states. A scope is currently active when its [node] is the
-/// first focus of its parent scope. To activate a [FocusScope], either use the
-/// [autofocus] property or explicitly make the [node] the first focus in the
-/// parent scope:
+/// It provides [onFocusChange] as a way to be notified when the focus is given
+/// to or removed from this widget.
 ///
-/// ```dart
-/// FocusScope.of(context).setFirstFocus(node);
-/// ```
+/// The [onKey] argument allows specification of a key even handler that should
+/// be invoked when this node or one of its children has focus.
 ///
-/// If a [FocusScope] is removed from the widget tree, then the previously
-/// focused node will be focused, but only if the [node] is the same [node]
-/// object as in the previous frame. To assure this, you can use a GlobalKey to
-/// keep the [FocusScope] widget from being rebuilt from one frame to the next,
-/// or pass in the [node] from a parent that is not rebuilt. If there is no next
-/// sibling, then the parent scope node will be focused.
+/// This widget does not provide any visual indication that the focus has
+/// changed. To provide that, add a [FocusHighlight] widget as a descendant of
+/// this widget.
 ///
-/// See also:
+/// To collect nodes into a group, use a [FocusScopeNode].
 ///
-///  * [FocusScopeNode], which is the associated node in the focus tree.
-///  * [FocusNode], which is a leaf node in the focus tree that can receive
-///    focus.
-class FocusScope extends StatefulWidget {
-  /// Creates a scope in which widgets can receive focus.
+/// To manipulate the focus, use methods on [FocusScopeNode]. For instance,
+/// to move the focus to the next node, call
+/// `Focus.of(context).nextFocus()`.
+class Focus extends StatefulWidget {
+  /// Creates a widget that manages a [FocusNode]
```

Comment:
punctuation

Reply:
Fixed. What is it about constructors that makes me do that? :-)

Reply:
constructors are a pain to document in general. dunno why they make you omit punctuation, but they certainly give me the most difficulty in terms of not writing useless docs.